org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/lookup/InferenceContext18.java

```diff
@@ -1910,4 +1784,412 @@ public void cleanUp() {
+						((IntersectionTypeBinding18) candidateT).getIntersectingTypes();
+			if (allBoundCandidates != null) {
+				for (TypeBinding t : allBoundCandidates) {
+					if (typeBinding.isCompatibleWith(t))
+						return true;
+				}
+			}
+			return false;
+		}
+		return typeBinding.isCompatibleWith(candidateT);
+	}
+
+	/* 6. => 18_5_5_item_6
+	 * Otherwise, let A1, ..., An be the resolved instantiations for α1, ..., αn, and
+	 * let Y1, ..., Yp (p ≥ 0) be any fresh type variables introduced by resolution.
+	 * The type of the record pattern is the upward projection of R<A1, ..., An>
+	 * with respect to Y1, ..., Yp (4.10.5).
+	 */
+	private ReferenceBinding getRecordPatternTypeFromUpwardsProjection(TypeBinding typeBinding,
+			InferenceVariable[] alphas, BoundSet solution) {
+		TypeBinding[] instantiations = new TypeBinding[alphas.length];
+		for (int i = 0, l = alphas.length; i < l; ++i) {
+			instantiations[i] = solution.getInstantiation(alphas[i], this.environment);
+		}
+		ReferenceBinding r = (ReferenceBinding) typeBinding.original();
+		ParameterizedTypeBinding rA = this.environment.createParameterizedType(
+				r,
+				instantiations,
+				r.enclosingType(),
+				r.getTypeAnnotations());
+		TypeVariableBinding[] yTypeVariables = getFreshTypeVariables(instantiations);
+//		if (yTypeVariables == null || yTypeVariables.length == 0)
+//			return null;
+
+		return rA != null ? yTypeVariables == null || yTypeVariables.length == 0 ? rA :
+				rA.upwardsProjection(this.scope, yTypeVariables) : null;
+	}
+
+	private TypeVariableBinding[] getFreshTypeVariables(TypeBinding[] instantiations) {
+		if (instantiations == null)
+			return null;
+		List<CaptureBinding> yTypeVariables = new ArrayList<>();
+		for (TypeBinding b : instantiations) {
+			// ...be any fresh type variables introduced by resolution. - how to get this condition?
+			if (b instanceof CaptureBinding)
+				yTypeVariables.add((CaptureBinding) b);
+		}
+		return yTypeVariables.toArray(new TypeVariableBinding[0]);
+	}
+
+	/* 4. => 18_5_5_item_4
+	 * If T' is a parameterization of a generic class G, and there exists a supertype of
+	 * R<α1, ..., αn> that is also a parameterization of G, let R' be that supertype */
+	private boolean findRPrimeAndResultingBounds(TypeBinding typeBinding, InferenceVariable[] alphas, TypeBinding tPrime) {
+		if (tPrime == null)
+			return false;
+		ReferenceBinding rAlpha = this.environment.createParameterizedType(
+				(ReferenceBinding) typeBinding.original(), alphas, typeBinding.enclosingType(), typeBinding.getTypeAnnotations());
+		TypeBinding[] rPrimes = this.currentBounds.condition18_5_5_item_4(rAlpha, alphas, tPrime, this);
+		if (rPrimes != null) {
+			// TODO:
+			/* The constraint formula ‹T' = R'› is reduced (18.2) and the resulting bounds are
+			 * incorporated into B1 to produce a new bound set, B2.
+			 * TODO: refactor the method below into two? - instead of creating array - here reusing. */
```

Comment:
Re "TODO: refactor...":

I think `reduceWithEqualityConstraints` is just not needed here, what about
```
try {
	if (!reduceAndIncorporate(ConstraintTypeFormula.create(tPrime, rPrime, ReductionResult.SAME))) {
		 /* If B2 contains the bound false, inference fails. */
		return false;
	}
} catch (InferenceFailureException e) {
	return false;
}
```